### ISSUE TITLE
The auto-borrowing system in the context of filters accepting params by value

### ISSUE THREAD

**Issue by couchand** [djc/askama#550]
After reading the conversations on #423 and #492 I'm convinced that filters that take `Copy` types should heceforth accept their parameters by value.  This makes filters very nicely line up with general API design of `Copy` types.

Passing function call return values and literals is exactly right: `{{ 1|abs }}` or `{{ list.len()|abs }}`.

There's just one thing: the very common case of passing template variables is now burdened by an apparently-unnecessary `.clone()`.  What was previously the perfectly clear `{{ field|filter }}` is now (with the API change to take by value) `{{ field.clone()|filter }}`.

I'm not sure what the right solution is: obviously the auto-borrow is very useful and necessary in most cases.  This wrinkle doesn't seem significant enough to justify introducing `*` and `&`.  I don't think codegen-time has visibility into trait impls.  Maybe all filters should just continue to accept their arguments by reference as a general practice?  (What of `abs`, then?)

**Comment by couchand** [comment]
I've thought a bit more about this and something occurred to me: in Rust, there is one and only one idiom for "this can take a owned or borrowed thing and you don't need to care which": method call syntax.

```rust
let owned = 42;
let borrowed = &owned;
assert_eq!(owned.abs(), borrowed.abs())
```

This kind of suggests that there just shouldn't be a filter like `| abs`, since the idiomatic Rust way to write the template would be to use `.abs()`.  I suppose the fact that it was never possible to use until now and nobody complained is a good indication that it's not pulling its weight anyway...  (See aside below)

So I've now completely reversed positions from my initial post -- I now think that all filters should take their (first) argument by reference.  But what about the extra parameters like on the filters `indent` and `truncate`?  Given how prevalent you'd expect literal usage with them (e.g. `{{ text | truncate(4) }}`) the subsequent arguments should probably be by value.

Now we have a weird mis-match, where the first argument of a filter (passed in through the pipe) is always by reference but subsequent arguments are by value.  When we're passing literals or function call results into a filter, we'd really like them to be auto-borrowed for us.  When we're passing `Copy` type template fields into function calls, we'd really like them to *not* be auto-borrowed.  That is, in this environment:

```rust
#[derive(Template)]
struct Example {
    num: isize,
    text: String,
    object: Object,
}

struct Object;
impl Object {
    fn method(&self) -> isize { todo!() }
    fn munge(&self, other: &isize) -> isize { todo!() }
}
```

We want to write:

```
{{ num | abs }}
{{ 1 | abs }}
{{ object.method() | abs }}
```

And also:

```
{{ text | truncate(num) }}
{{ text | truncate(1) }}
{{ text | truncate(object.method()) }}
```

And also:

```
{% for i in items.iter().skip(num) %}{{ i }}{% endfor %}
{% for i in items.iter().skip(1) %}{{ i }}{% endfor %}
{% for i in items.iter().skip(object.method()) %}{{ i }}{% endfor %}
```

And I believe this system then implies that function calls that accept reference are now:

```
{{ object.munge(num.borrow()) }}
{{ object.munge(1isize.borrow()) }}
{{ object.munge(object.method.borrow()) }}
```

This all leads me to believe that the thing which determines whether we should be auto-borrowing is not whether it's a literal, call, or template field, but rather, how you're using the value.

---

Aside: should the filter syntax codegen to method call syntax?  They kind of already look like methods if you squint.  You could use any method in scope, and implement custom filters as traits.  So Rusty!  My oh my that would be a huge breaking change, though.

**Comment by msrd0** [comment]
> in Rust, there is one and only one idiom for "this can take an owned or borrowed thing and you don't need to care which": method call syntax.

This is not true. There is a second one directly in the standard library, namely all macros that work with format, like `format!`, `print!`, `write!` etc. In a function call, this won't work, as the function invokation "eats" the variable that is not `Copy`:

```rust
let s: String = "foo".to_owned();
println("{}", s);
assert_eq!(s.as_str(), "foo"); // s cannot be used anymore
```

However, using macros from the standard library, this works, and further, also if we take a reference:

```rust
let s: String = "foo".to_owned();
println!("{}", s); // works, doesn't consume s
println!("{}", &s); // does exactly the same as above
assert_eq!(s.as_str(), "foo"); // s can still be used
```

Since askama is a macro, it is IMHO not only idiomatic to have it accept both references and owned values, but also in line with the macros implemented in the standard library.

**Comment by couchand** [comment]
Interesting angle, @msrd0.  Though askama is a macro, it emits a trait implementation, so the code it's lowered to has to play by the rules of functions, not macros.  I think the discussion above suggests that there's not really a way to emit code that will handle the distinction without intervention on the part of the template author.  If you've got a suggestion of how to implement the more flexible behavior please mention it!

**Comment by msrd0** [comment]
Sure. Just looking at the expansion of above code, I was able to reproduce the idea behind it ([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b4e57ea392cf1a5de44822b46c8a32c1)):

```rust
fn strlen(s: &str) -> usize {
    s.len()
}

macro_rules! strlen {
    ($s:expr) => {
        match (&$s,) {
            s => strlen(s.0),
        }
    };
}

fn main() {
    let s: String = "foo".to_owned();
    println!("{}", strlen!(s));
    println!("{}", strlen!(&s));
}
```

So askama could (if it had a filter `strlen`, which it doesn't, but was an easy example for my proof of concept) emit the code that above macro emits instead of calling the `strlen` function directly, and thereby work with both references and owned values.

**Comment by couchand** [comment]
That's a nice idea @msrd0, and seems to go to the idea that all filters should be written to take their argument by reference (since that pattern will not compile if the function takes a value).

---

On another note, how about the result of a filter?  Looking into the code, I noticed that after #423, this no longer compiles: `{{ foo|truncate(bar|abs) }}`.  Perhaps we should remove the auto-borrow from a filter result, since it's roughly equivalent to a function call?
